File: src/gtk/secret_store.zig
//! Secret Service/libsecret implementation of relay_core's CredStore.

const std = @import("std");
const relay = @import("relay_core");

const store = relay.cred.store;
const Diagnostics = relay.diag.Diagnostics;

pub const SecretStore = struct {
    pub fn credStore(self: *SecretStore) store.CredStore {
        return .{ .ctx = @ptrCast(self), .vtable = &vtable };
    }

    fn get(
        _: *anyopaque,
        gpa: std.mem.Allocator,
        diag: *Diagnostics,
        key: store.Key,
    ) store.Error![]u8 {
        var secret: ?[*:0]u8 = null;
        var secret_len: usize = 0;
        var message: ?[*:0]u8 = null;
        const result = relay_secret_lookup(
            protocolName(key.protocol).ptr,
            protocolName(key.protocol).len,
            key.host.ptr,
            key.host.len,
            key.port,
            key.account.ptr,
            key.account.len,
            &secret,
            &secret_len,
            &message,
        );
        defer freeMessage(message);
        switch (result) {
            .ok => {},
            .not_found => return error.NotFound,
            .access_denied => {
                setDiag(diag, .auth, message, "Secret Service denied credential access");
                return error.AccessDenied;
            },
            .unexpected => {
                setDiag(diag, .permanent, message, "Secret Service credential lookup failed");
                return error.Unexpected;
            },
        }
        const value = secret orelse return error.Unexpected;
        defer relay_secret_password_free(value);
        return gpa.dupe(u8, value[0..secret_len]);
    }

    fn set(
        _: *anyopaque,
        diag: *Diagnostics,
        key: store.Key,
        secret: []const u8,
    ) store.Error!void {
        var message: ?[*:0]u8 = null;
        const result = relay_secret_store(
            protocolName(key.protocol).ptr,
            protocolName(key.protocol).len,
            key.host.ptr,
            key.host.len,
            key.port,
            key.account.ptr,
            key.account.len,
            secret.ptr,
            secret.len,
            &message,
        );
        defer freeMessage(message);
        return resultError(result, diag, message, "Secret Service credential write failed");
    }

    fn delete(_: *anyopaque, diag: *Diagnostics, key: store.Key) store.Error!void {
        var message: ?[*:0]u8 = null;
        const result = relay_secret_clear(
            protocolName(key.protocol).ptr,
            protocolName(key.protocol).len,
            key.host.ptr,
            key.host.len,
            key.port,
            key.account.ptr,
            key.account.len,
            &message,
        );
        defer freeMessage(message);
        return resultError(result, diag, message, "Secret Service credential deletion failed");
    }

    const vtable: store.VTable = .{
        .get = get,
        .set = set,
        .delete = delete,
    };
};

const Result = enum(c_int) {
    ok = 0,
    not_found = 1,
    access_denied = 2,
    unexpected = 3,
};

fn protocolName(protocol: store.Protocol) []const u8 {
    return @tagName(protocol);
}

fn resultError(result: Result, diag: *Diagnostics, message: ?[*:0]u8, fallback: []const u8) store.Error!void {
    switch (result) {
        .ok, .not_found => return,
        .access_denied => {
            setDiag(diag, .auth, message, fallback);
            return error.AccessDenied;
        },
        .unexpected => {
            setDiag(diag, .permanent, message, fallback);
            return error.Unexpected;
        },
    }
}

fn setDiag(
    diag: *Diagnostics,
    class: relay.diag.ErrorClass,
    message: ?[*:0]u8,
    fallback: []const u8,
) void {
    const text = if (message) |value| std.mem.span(value) else fallback;
    diag.set(class, 0, "{s}", .{text});
}

fn freeMessage(message: ?[*:0]u8) void {
    if (message) |value| relay_secret_message_free(value);
}

extern fn relay_secret_lookup(
    protocol: [*]const u8,
    protocol_len: usize,
    host: [*]const u8,
    host_len: usize,
    port: u16,
    account: [*]const u8,
    account_len: usize,
    secret_out: *?[*:0]u8,
    secret_len_out: *usize,
    message_out: *?[*:0]u8,
) Result;

extern fn relay_secret_store(
    protocol: [*]const u8,
    protocol_len: usize,
    host: [*]const u8,
    host_len: usize,
    port: u16,
    account: [*]const u8,
    account_len: usize,
    secret: [*]const u8,
    secret_len: usize,
    message_out: *?[*:0]u8,
) Result;

extern fn relay_secret_clear(
    protocol: [*]const u8,
    protocol_len: usize,
    host: [*]const u8,
    host_len: usize,
    port: u16,
    account: [*]const u8,
    account_len: usize,
    message_out: *?[*:0]u8,
) Result;

extern fn relay_secret_password_free(secret: [*:0]u8) void;
extern fn relay_secret_message_free(message: [*:0]u8) void;

test "CredStore vtable is available without touching the user's keyring" {
    var backend: SecretStore = .{};
    const cred_store = backend.credStore();
    try std.testing.expect(cred_store.ctx == @as(*anyopaque, @ptrCast(&backend)));
}

test {
    std.testing.refAllDecls(SecretStore);
}

File: src/gtk/secret_store.c
#include <libsecret/secret.h>
#include <gio/gio.h>
#include <stddef.h>
#include <stdint.h>
#include <string.h>

enum relay_secret_result {
    RELAY_SECRET_OK = 0,
    RELAY_SECRET_NOT_FOUND = 1,
    RELAY_SECRET_ACCESS_DENIED = 2,
    RELAY_SECRET_UNEXPECTED = 3,
};

static const SecretSchema relay_schema = {
    "us.doriantull.relay.credentials",
    SECRET_SCHEMA_NONE,
    {
        { "protocol", SECRET_SCHEMA_ATTRIBUTE_STRING },
        { "host", SECRET_SCHEMA_ATTRIBUTE_STRING },
        { "port", SECRET_SCHEMA_ATTRIBUTE_STRING },
        { "account", SECRET_SCHEMA_ATTRIBUTE_STRING },
        { NULL, 0 },
    },
};

static enum relay_secret_result relay_error_result(GError *error, char **message_out) {
    if (message_out != NULL) {
        *message_out = g_strdup(error != NULL && error->message != NULL
            ? error->message
            : "Secret Service request failed");
    }
    enum relay_secret_result result = RELAY_SECRET_UNEXPECTED;
    if (error != NULL && (
        g_error_matches(error, SECRET_ERROR, SECRET_ERROR_IS_LOCKED) ||
        g_error_matches(error, G_IO_ERROR, G_IO_ERROR_PERMISSION_DENIED) ||
        g_error_matches(error, G_DBUS_ERROR, G_DBUS_ERROR_ACCESS_DENIED) ||
        g_error_matches(error, G_DBUS_ERROR, G_DBUS_ERROR_AUTH_FAILED)
    )) {
        result = RELAY_SECRET_ACCESS_DENIED;
    }
    if (error != NULL) g_error_free(error);
    return result;
}

static char *relay_dupe(const uint8_t *bytes, size_t len) {
    if (memchr(bytes, '\0', len) != NULL) return NULL;
    return g_strndup((const char *)bytes, len);
}

int relay_secret_lookup(
    const uint8_t *protocol, size_t protocol_len,
    const uint8_t *host, size_t host_len,
    uint16_t port,
    const uint8_t *account, size_t account_len,
    char **secret_out, size_t *secret_len_out, char **message_out
) {
    GError *error = NULL;
    char *protocol_z = relay_dupe(protocol, protocol_len);
    char *host_z = relay_dupe(host, host_len);
    char *account_z = relay_dupe(account, account_len);
    char port_z[6];
    g_snprintf(port_z, sizeof(port_z), "%u", (unsigned)port);
    if (protocol_z == NULL || host_z == NULL || account_z == NULL) {
        g_free(protocol_z); g_free(host_z); g_free(account_z);
        return RELAY_SECRET_UNEXPECTED;
    }
    char *secret = secret_password_lookup_sync(
        &relay_schema, NULL, &error,
        "protocol", protocol_z,
        "host", host_z,
        "port", port_z,
        "account", account_z,
        NULL
    );
    g_free(protocol_z); g_free(host_z); g_free(account_z);
    if (error != NULL) return relay_error_result(error, message_out);
    if (secret == NULL) return RELAY_SECRET_NOT_FOUND;
    *secret_out = secret;
    *secret_len_out = strlen(secret);
    return RELAY_SECRET_OK;
}

int relay_secret_store(
    const uint8_t *protocol, size_t protocol_len,
    const uint8_t *host, size_t host_len,
    uint16_t port,
    const uint8_t *account, size_t account_len,
    const uint8_t *secret, size_t secret_len,
    char **message_out
) {
    GError *error = NULL;
    char *protocol_z = relay_dupe(protocol, protocol_len);
    char *host_z = relay_dupe(host, host_len);
    char *account_z = relay_dupe(account, account_len);
    char *secret_z = relay_dupe(secret, secret_len);
    char port_z[6];
    g_snprintf(port_z, sizeof(port_z), "%u", (unsigned)port);
    if (protocol_z == NULL || host_z == NULL || account_z == NULL || secret_z == NULL) {
        g_free(protocol_z); g_free(host_z); g_free(account_z); g_free(secret_z);
        return RELAY_SECRET_UNEXPECTED;
    }
    gboolean ok = secret_password_store_sync(
        &relay_schema, SECRET_COLLECTION_DEFAULT, "Relay server password",
        secret_z, NULL, &error,
        "protocol", protocol_z,
        "host", host_z,
        "port", port_z,
        "account", account_z,
        NULL
    );
    secret_password_wipe(secret_z);
    g_free(secret_z);
    g_free(protocol_z); g_free(host_z); g_free(account_z);
    if (error != NULL) return relay_error_result(error, message_out);
    if (!ok) return RELAY_SECRET_UNEXPECTED;
    return RELAY_SECRET_OK;
}

int relay_secret_clear(
    const uint8_t *protocol, size_t protocol_len,
    const uint8_t *host, size_t host_len,
    uint16_t port,
    const uint8_t *account, size_t account_len,
    char **message_out
) {
    GError *error = NULL;
    char *protocol_z = relay_dupe(protocol, protocol_len);
    char *host_z = relay_dupe(host, host_len);
    char *account_z = relay_dupe(account, account_len);
    char port_z[6];
    g_snprintf(port_z, sizeof(port_z), "%u", (unsigned)port);
    if (protocol_z == NULL || host_z == NULL || account_z == NULL) {
        g_free(protocol_z); g_free(host_z); g_free(account_z);
        return RELAY_SECRET_UNEXPECTED;
    }
    gboolean ok = secret_password_clear_sync(
        &relay_schema, NULL, &error,
        "protocol", protocol_z,
        "host", host_z,
        "port", port_z,
        "account", account_z,
        NULL
    );
    g_free(protocol_z); g_free(host_z); g_free(account_z);
    if (error != NULL) return relay_error_result(error, message_out);
    if (!ok) return RELAY_SECRET_NOT_FOUND;
    return RELAY_SECRET_OK;
}

void relay_secret_password_free(char *secret) {
    secret_password_free(secret);
}

void relay_secret_message_free(char *message) {
    g_free(message);
}
